content/blog/callpackage-a-tool-for-the-lazy.md

```diff
@@ -0,0 +1,208 @@
+---
+title: callPackage, a tool for the lazy
+date: 2022-08-25
+extra:
+  author: Norbert Melzer
+---
+
+The Nix package manager ships with a a special-purpose programming language for creating packages and configurations: the Nix language.
+It is used to build the Nix package collection, known as [`nixpkgs`](https://github.com/nixos/nixpkgs) – the largest, most up-to-date open source software distribution in the world.
+
+Being purely functional, the Nix language allows declaring custom functions to abstract over common patterns.
+One such pattern is parametrization of package definitions, that is, builds which can vary by their dependencies or settings.
+
+`nixpkgs` is a sizeable software project on it's own, with coding conventions and idioms that have emerged over the years.
+It has [established a convention](https://github.com/NixOS/nixpkgs/pull/9869) of composing parameterized packages with automatic settings through a function named [`callPackage`](https://github.com/NixOS/nixpkgs/commit/fd268b4852d39c18e604c584dd49a611dc795a9b).
+
+This article shows how to use it and why it's beneficial.
+
+## Basic examples
+
+Before even discussing the benefits, let's see how it actually gets used.
+
+Given are the files `hello.nix` and `default.nix`:
+
+```nix
+# default.nix
+let pkgs = import <nixpkgs> {}; in
+pkgs.callPackage ./hello.nix {}
+```
+
+`default.nix` produces a *derivation* from the contents of `hello.nix`.
+This what the Nix package manager calls its build results (which are often packaged executables, but can be arbitrary files), and what the Nix language calls the data type of such an expression.
```

Comment:
```suggestion
This is what the Nix package manager calls its build results (which are often packaged executables, but can be arbitrary files), and what the Nix language calls the data type of such an expression.
```

Is "This" referring to *derivation*? And is *build results* an official term? I guess we are skipping the whole .drv file story here if you say it's executables and files. Maybe that's good but we could try to at least mention it so that we don't confuse those who know about them.

Reply:
"build result" is something we came up with with @Ericson2314 when trying to explain Nix architecture with more generic terms. Actually that original sentence turned out wrong, because derivations (which we generically called "build task") are descriptions of how to obtain build results.

```suggestion
A derivation is what the Nix package manager calls a description of how to obtain a desired build result. Build results are often packaged executables, but can be arbitrary files. Derivation is also the data type of the corresponding expression in the Nix language.
```